crates/wasm/src/view_server.rs

```diff
@@ -91,112 +92,30 @@ impl ViewServer {
+    /// Use `get_updates()` to get the scan results
+    /// Returns: `bool`
     #[wasm_bindgen]
-    pub fn get_updates(
-        &mut self,
-        last_position: JsValue,
-        last_forgotten: JsValue,
-    ) -> Result<JsValue, Error> {
-        let result = self.get_updates_inner(last_position, last_forgotten)?;
-        serde_wasm_bindgen::to_value(&result)
-    }
-
-    /// get SCT root
-    /// SCT root can be compared with the root obtained by GRPC and verify that there is no divergence
-    /// Returns: `Root`
-    #[wasm_bindgen]
-    pub fn get_nct_root(&mut self) -> Result<JsValue, Error> {
-        let root = self.nct.root();
-        serde_wasm_bindgen::to_value(&root)
-    }
-
-    /// get LP NFT asset
-    /// Arguments:
-    ///     position_value: `lp::position::Position`
-    ///     position_state_value: `lp::position::State`
-    /// Returns: `DenomMetadata`
-    #[wasm_bindgen]
-    pub fn get_lpnft_asset(
-        &mut self,
-        position_value: JsValue,
-        position_state_value: JsValue,
-    ) -> Result<JsValue, Error> {
-        let position: Position = serde_wasm_bindgen::from_value(position_value)?;
-        let position_state = serde_wasm_bindgen::from_value(position_state_value)?;
-        let lp_nft = LpNft::new(position.id(), position_state);
-        let denom = lp_nft.denom();
-        serde_wasm_bindgen::to_value(&denom)
-    }
-}
-
-impl ViewServer {
-    pub fn get_updates_inner(
-        &mut self,
-        last_position: JsValue,
-        last_forgotten: JsValue,
-    ) -> WasmResult<ScanBlockResult> {
-        let stored_position: Option<StoredPosition> =
-            serde_wasm_bindgen::from_value(last_position)?;
-        let stored_forgotten: Option<Forgotten> = serde_wasm_bindgen::from_value(last_forgotten)?;
-
-        let nct_updates: Updates = self
-            .nct
-            .updates(
-                stored_position.unwrap_or_default(),
-                stored_forgotten.unwrap_or_default(),
-            )
-            .collect::<Updates>();
-
-        let result = ScanBlockResult {
-            height: self.latest_height,
-            nct_updates,
-            new_notes: self.notes.clone().into_values().collect(),
-            new_swaps: self.swaps.clone().into_values().collect(),
-        };
-        Ok(result)
-    }
-
-    pub async fn scan_block_inner(
-        &mut self,
-        compact_block: JsValue,
-    ) -> WasmResult<ScanBlockResult> {
+    pub async fn scan_block(&mut self, compact_block: JsValue) -> WasmResult<bool> {
         let block_proto: penumbra_proto::core::component::compact_block::v1alpha1::CompactBlock =
             serde_wasm_bindgen::from_value(compact_block)?;
 
         let block: CompactBlock = block_proto.try_into()?;
 
-        // Newly detected spendable notes.
-        let mut new_notes = Vec::new();
-        // Newly detected claimable swaps.
-        let mut new_swaps: Vec<SwapRecord> = Vec::new();
+        let mut found_new_data: bool = false;
```

Comment:
Do you see a usecase for this in the web code? I don't think this will be used.

Reply:
Yes, of course.  
I changed the web repo code to test the new wasm logic. 
Most of the tests is to compare the state of `indexed_db` and` pcli view db` for different cases. 

Regarding `found_new_data`, this is an important part of block scanning. If we don't handle `found_new_data` values, we run the risk of missing nullifier and not marking some notes as spent.

I did it this way 
```typescript
 const found_new_data: boolean = await this.viewServer.scanBlock(res.compactBlock);

      await this.markNotesSpent(res.compactBlock.nullifiers, res.compactBlock.height);

      if (isDevEnv()) {
        await this.assertRootValid(res.compactBlock.height);
      }

      if (found_new_data || shouldStoreProgress(res.compactBlock, lastBlockHeight)) {
        await this.saveSyncProgress(res.compactBlock.height);
      }
``` 

And moved the saving of new notes to `saveSyncProgress()`
```typescript
 async saveSyncProgress(height: bigint) {
    const updates = await this.viewServer.updatesSinceCheckpoint();
    await this.storeNewNotes(updates.new_notes);
    await this.indexedDb.updateStateCommitmentTree(updates.nct_updates, height);
  }
``` 
Save `notes` and `nct_updates` also need to be saved in the same transaction, but for my tests this wasn't critical, so I left it out.

Reply:
That means our db writes have that one exception (nullifiers) and everything else gets saved in a single transaction every 1000 blocks.

Could we just collect nullifiers as we find them and put them in memory on the struct? And then when we call `flush_updates()` we get those back/clears as well? That will allow us to handle all db writes only then.

Reply:
Also, do those `nullifiers` belong to anyone else? Is it guaranteed to belong to the user? If it is specifically for the user, I see that it makes sense to have `found_new_data`. If it contains public info, then I'd say it doesn't make sense for `found_new_data` to be triggered by nullifiers being present.

Reply:
There's a little miscommunication 
`nullifiers` are not processed in wasm and do not affect the `found_new_data` value in any way
`found_new_data = true` only when we found our notes or swaps when scanning the block 
`Nullifiers` should be checked on TS side, comparing `nullifiers` that are in block and `nullifiers` that we have in indexed_db. 
In 99.99% of cases this requires only reading from `indexed_db`, because we will not find nullifiers and we will not need to change data in `indexed_db` (mark `note` as spent).
Marking notes as spent doesn't affect SCT  state, so I don't see anything wrong with having it in a separate transaction

Reply:
Ah yes, that's right. I misunderstood. Ok, I think all looks good then.